lib/matplotlib/image.py

```diff
@@ -1052,14 +1052,51 @@ def make_image(self, renderer, magnification=1.0, unsampled=False):
             self._is_grayscale = False
         vl = self.axes.viewLim
         l, b, r, t = self.axes.bbox.extents
-        width = (round(r) + 0.5) - (round(l) - 0.5)
-        height = (round(t) + 0.5) - (round(b) - 0.5)
-        width *= magnification
-        height *= magnification
-        im = _image.pcolor(self._Ax, self._Ay, A,
-                           int(height), int(width),
-                           (vl.x0, vl.x1, vl.y0, vl.y1),
-                           _interpd_[self._interpolation])
+        width = int(((round(r) + 0.5) - (round(l) - 0.5)) * magnification)
+        height = int(((round(t) + 0.5) - (round(b) - 0.5)) * magnification)
+        x_pix = np.linspace(vl.x0, vl.x1, width)
+        y_pix = np.linspace(vl.y0, vl.y1, height)
+        if self._interpolation == "nearest":
+            x_mid = (self._Ax[:-1] + self._Ax[1:]) / 2
+            y_mid = (self._Ay[:-1] + self._Ay[1:]) / 2
+            x_int = x_mid.searchsorted(x_pix)
+            y_int = y_mid.searchsorted(y_pix)
+            # The following is equal to `A[y_int[:, None], x_int[None, :]]`,
+            # but many times faster.  Both casting to uint32 (to have an
+            # effectively 1D array) and manual index flattening matter.
+            im = (
+                np.ascontiguousarray(A).view(np.uint32).ravel()[
+                    np.add.outer(y_int * A.shape[1], x_int)]
+                .view(np.uint8).reshape((height, width, 4)))
+        else:  # self._interpolation == "bilinear"
+            # Use np.interp to compute x_int/x_float has similar speed.
```

Comment:
You say that `np.interp` is approximately equally fast, but still decide to implement the interpolation yourself? Why?

If we want our own interpolation, I'd still favor a dedicated private function. That would make it more clear and simpler to test and profile.

Reply:
Because in this specific case you also need to tweak a bit `interp`, i.e. the actual implementation is something like (from memory)
```python
            range_ax = np.arange(len(self._Ax))
            # Don't index beyond the end.
            range_ax[-1] = np.nextafter(len(self._Ax) - 1, 0)
            x = np.interp(x_pix, self._Ax, range_ax)
            x_int = x.astype(int)
            x_frac = np.modulo(x, np.float32)
```
which I don't think is more readable (it's not really worse either).

I'm not convinced factoring this out into e.g. `_interpolate(A, self._Ax, self._Ay, x_pix, y_pix)` would help legibility either.

Also, re profiling, the real bottleneck is not actually here, it's in the actual interpolation code below.